File: crates/catalog/src/projections/family_detail.rs
use crate::errors::{CatalogError, CatalogResult};
use serde_json::{json, Value};
use sqlx::{PgPool, Row};

pub async fn rebuild_family_detail(
    pool: &PgPool,
    tenant_id: &str,
) -> CatalogResult<Vec<(String, Value)>> {
    let families = sqlx::query(
        r#"
        SELECT id, key, family_type, name, canonical_slug, current_revision_number, published_revision_number
        FROM catalog_families WHERE tenant_id = $1
        "#,
    )
    .bind(tenant_id)
    .fetch_all(pool)
    .await
    .map_err(|e| CatalogError::StorageFailed(e.to_string()))?;

    let mut results = Vec::new();

    for fam in families {
        let family_id: uuid::Uuid = fam.get("id");
        let family_key: String = fam.get("key");
        let family_type: String = fam.get("family_type");
        let name: String = fam.get("name");
        let slug: String = fam.get("canonical_slug");
        let current_rev: i32 = fam.get("current_revision_number");
        let published_rev: Option<i32> = fam.get("published_revision_number");

        let attrs = sqlx::query(
            r#"
            SELECT a.key AS attr_key, a.data_type, fa.role, fa.required, fa.filterable,
                   fa.sortable, fa.is_variant_axis, fa.display_order
            FROM catalog_family_attributes fa
            JOIN catalog_attribute_definitions a ON a.id = fa.attribute_id
            WHERE fa.family_id = $1
            ORDER BY fa.display_order
            "#,
        )
        .bind(family_id)
        .fetch_all(pool)
        .await
        .map_err(|e| CatalogError::StorageFailed(e.to_string()))?;

        let attr_list: Vec<Value> = attrs
            .iter()
            .map(|r| {
                json!({
                    "attributeKey": r.get::<String, _>("attr_key"),
                    "dataType": r.get::<String, _>("data_type"),
                    "role": r.get::<String, _>("role"),
                    "required": r.get::<bool, _>("required"),
                    "filterable": r.get::<bool, _>("filterable"),
                    "sortable": r.get::<bool, _>("sortable"),
                    "isVariantAxis": r.get::<bool, _>("is_variant_axis"),
                    "displayOrder": r.get::<i32, _>("display_order"),
                })
            })
            .collect();

        let display_policies = sqlx::query(
            r#"
            SELECT dp.surface, a.key AS attr_key, dp.role, dp.display_order
            FROM catalog_family_display_policies dp
            JOIN catalog_attribute_definitions a ON a.id = dp.attribute_id
            WHERE dp.family_id = $1
            ORDER BY dp.surface, dp.display_order
            "#,
        )
        .bind(family_id)
        .fetch_all(pool)
        .await
        .map_err(|e| CatalogError::StorageFailed(e.to_string()))?;

        let dp_list: Vec<Value> = display_policies
            .iter()
            .map(|r| {
                json!({
                    "surface": r.get::<String, _>("surface"),
                    "attributeKey": r.get::<String, _>("attr_key"),
                    "role": r.get::<String, _>("role"),
                    "order": r.get::<i32, _>("display_order"),
                })
            })
            .collect();

        let assets = sqlx::query(
            r#"
            SELECT a.asset_key, a.media_type, a.uri, aa.role, aa.display_order
            FROM catalog_asset_attachments aa
            JOIN catalog_assets a ON a.id = aa.asset_id
            WHERE aa.family_id = $1
            ORDER BY aa.display_order
            "#,
        )
        .bind(family_id)
        .fetch_all(pool)
        .await
        .map_err(|e| CatalogError::StorageFailed(e.to_string()))?;

        let asset_list: Vec<Value> = assets
            .iter()
            .map(|r| {
                json!({
                    "assetKey": r.get::<String, _>("asset_key"),
                    "mediaType": r.get::<Option<String>, _>("media_type"),
                    "uri": r.get::<Option<String>, _>("uri"),
                    "role": r.get::<String, _>("role"),
                    "order": r.get::<i32, _>("display_order"),
                })
            })
            .collect();

        let payload = json!({
            "familyId": family_id.to_string(),
            "familyKey": family_key.clone(),
            "type": family_type,
            "name": name,
            "canonicalSlug": slug,
            "currentRevision": current_rev,
            "publishedRevision": published_rev,
            "attributes": attr_list,
            "displayPolicies": dp_list,
            "assets": asset_list,
        });

        sqlx::query(
            r#"
            INSERT INTO catalog_family_detail_projection (tenant_id, family_key, family_id, payload, updated_at)
            VALUES ($1, $2, $3, $4, NOW())
            ON CONFLICT (tenant_id, family_key) DO UPDATE SET
                family_id = EXCLUDED.family_id,
                payload = EXCLUDED.payload,
                updated_at = NOW()
            "#,
        )
        .bind(tenant_id)
        .bind(&family_key)
        .bind(family_id)
        .bind(&payload)
        .execute(pool)
        .await
        .map_err(|e| CatalogError::StorageFailed(e.to_string()))?;

        results.push((family_key, payload));
    }

    Ok(results)
}

File: packages/sim/src/domain/catalog/projections/family-detail.ts
import type { Db } from '../../../ports/db.ts';
import type { ProjectionStorePort } from '../../../ports/projection-store.ts';
import type { SeedPayload } from '../seed-types.ts';
import { deterministicUuid } from '../ids.ts';

export function projectionKey(familyKey: string, tenantId: string): string {
  return `catalog:family-detail:${familyKey}:${tenantId}`;
}

export async function rebuildFamilyDetail(
  tenantId: string,
  db: Db,
  projections: ProjectionStorePort,
): Promise<Array<{ familyKey: string; payload: unknown }>> {
  const state = await db.get('catalog_state', tenantId);
  if (!state) return [];
  const seed = state.payload as SeedPayload;
  const attrTypeByKey = new Map<string, string>();
  for (const a of seed.attributeDefinitions ?? []) {
    attrTypeByKey.set(a.key, a.dataType);
  }
  const publishedRevisions = state.publishedRevisions;

  const out: Array<{ familyKey: string; payload: unknown }> = [];
  for (const fam of seed.families) {
    const familyId = deterministicUuid('family', tenantId, fam.key);
    const sortedAttrs = [...fam.attributes].sort((a, b) => a.displayOrder - b.displayOrder);
    const attributes = sortedAttrs.map((fa) => ({
      attributeKey: fa.attributeKey,
      dataType: attrTypeByKey.get(fa.attributeKey) ?? 'string',
      role: fa.role,
      required: fa.required ?? false,
      filterable: fa.filterable ?? false,
      sortable: fa.sortable ?? false,
      isVariantAxis: fa.isVariantAxis ?? false,
      displayOrder: fa.displayOrder,
    }));
    const dps = (fam.displayPolicies ?? [])
      .slice()
      .sort((a, b) => a.surface.localeCompare(b.surface) || a.order - b.order)
      .map((dp) => ({
        surface: dp.surface,
        attributeKey: dp.attributeKey,
        role: dp.role,
        order: dp.order,
      }));

    const payload = {
      familyId,
      familyKey: fam.key,
      type: fam.type,
      name: fam.name,
      canonicalSlug: fam.canonicalSlug,
      currentRevision: 1,
      publishedRevision: publishedRevisions[fam.key] ?? null,
      attributes,
      displayPolicies: dps,
      assets: [] as unknown[],
    };
    await projections.set(projectionKey(fam.key, tenantId), payload);
    out.push({ familyKey: fam.key, payload });
  }
  return out;
}
